### ISSUE TITLE
Crashed at Inner::try_wake()

### ISSUE THREAD

**Issue by black-binary** [Mnwa/fast-async-mutex#2]
My code crashed after switching from smol::lock::Mutex to fast-async-mutex::mutex::Mutex. I'm not familiar with the unsafe code so I'm not sure if it's a bug or I'm using it in the wrong way.

https://github.com/Mnwa/fast-async-mutex/blob/817788ecd44fdaf2b94e73382898defa020c9fb9/src/inner.rs#L98-L109

I need to lock the mutex in a non-async function. So what I'm doing is 

* calling lock.lock_owned() to generate a future `fut`

* calling `fut.poll(cx)` trying to lock the mutex immediately

* if poll() returns Pending, store the future in my struct, and wait for the next polling

Here is the original code

https://github.com/black-binary/ap-kcp/blob/ce0605523be63e8e68176847fc63c59a54b8918b/src/async_kcp.rs#L46-L73

And it crashed in the benchmark. I failed to reproduce this crash in debug mode, so I had to use gdb and IDA to find what happened.

The calling stack looks like this. The RIP register was zero (called a null pointer?)

![bt](https://user-images.githubusercontent.com/11072136/101275594-a3f47700-379e-11eb-92f5-d8d680cc5002.png)

IDA shows that the crash happened after calling Inner::try_wake()

![disasm](https://user-images.githubusercontent.com/11072136/101275703-9d1a3400-379f-11eb-9b9a-bec4852fe72f.png)

**Comment by Mnwa** [comment]
Hey! Thank you for this issue. Can you try to connect this lib from the `fix-try_wake` branch?
Like are
```toml
fast-async-mutex = { git = "https://github.com/Mnwa/fast-async-mutex", branch = "fix-try_wake" }
```
in your Cargo.toml
And retry you benchmark

**Comment by black-binary** [comment]
Thanks for the fix and the reply. Everything goes well after updating to that branch.

Looking forward to the new version. :)

**Comment by black-binary** [comment]
But the patch looks a little bit weird. Is the waker simply get deleted?

**Comment by Mnwa** [comment]
Nice! I publish the new version 0.6.7 with this fix.
Now `Weaker` will wake a future on every call of the `poll` method. It does not affect the performance a lot. I updated the Benchmark results in the Readme.

**Comment by Mnwa** [comment]
And now there is a much less unsafe code.

**Comment by black-binary** [comment]
> 
> 
> Nice! I publish the new version 0.6.7 with this fix.
> Now `Weaker` will wake a future on every call of the `poll` method. It does not affect the performance a lot. I updated the Benchmark results in the Readme.

I think it's not the right way to fix this. Every task calling to `poll` will be woken immediately. If a task locked the mutex for a long time, other tasks will keep polling it until the mutex is released (busy waiting). That sounds like a spinning lock instead of a mutex.

**Comment by Mnwa** [comment]
You're right. This is a fast fix of the bug. Next I will think about how I can do calling that on every poll.
